lib/internal/mime.js

```diff
@@ -16,62 +16,140 @@ const {
+
+    // #5-4 Advance position by 1.
+    index++;
+
+    // #5-5 If quoteOrBackslash is U+005C (\), then:
+    if (quoteOrBackslashAsCode === BACKSLASH_CHARCODE) {
+      // #5-5-1 If position is past the end of input, then append U+005C (\) to value and break.
+      if (index >= limit) {
+        value += BACKSLASH;
+        break;
+      }
+
+      // #5-5-2 Append the code point at position within input to value.
+      value += str[index];
+
+      // #5-5-3 Advance position by 1.
+      index++;
+    } else if (quoteOrBackslashAsCode === QUOTE_CHARCODE) {
+      // #5-6 Otherwise
+      // #5-6-1 Assert: quoteOrBackslash is U+0022 (").
+      // #5-6-2 Break
+      break;
+    }
+  }
+
+  marker.position = index;
+
+  // #6 If the extract-value flag is set, then return value
+  return value;
+}
+
 function parseTypeAndSubtype(str) {
-  // Skip only HTTP whitespace from start
-  let position = SafeStringPrototypeSearch(str, END_BEGINNING_WHITESPACE);
-  // read until '/'
-  const typeEnd = StringPrototypeIndexOf(str, SOLIDUS, position);
-  const trimmedType = typeEnd === -1 ?
-    StringPrototypeSlice(str, position) :
+  const strLength = str.length;
+  // #2 Let position be a position variable for input, initially pointg at
+  // the start of the input
+  let position = 0;
+
+  // #3 Let type be the result of collecting a sequence of code points that are
+  // not U+002F (/ [SOLIDUS]) from input, given position.
+  const typeEnd = StringPrototypeIndexOf(str, SOLIDUS);
+  const preType = typeEnd === -1 ?
+    str :
     StringPrototypeSlice(str, position, typeEnd);
-  const invalidTypeIndex = SafeStringPrototypeSearch(trimmedType,
-                                                     NOT_HTTP_TOKEN_CODE_POINT);
-  if (trimmedType === '' || invalidTypeIndex !== -1 || typeEnd === -1) {
-    throw new ERR_INVALID_MIME_SYNTAX('type', str, invalidTypeIndex);
+
+  // #4 If type is the empty string or does not solely
+  // contain HTTP token code points, then return failure.
+  // #5 If position is past the end of input, then return failure.
+  if (typeEnd === -1 || preType.length === 0 || typeEnd > strLength || SafeStringPrototypeSearch(
+    preType, NOT_HTTP_TOKEN_CODE_POINT) !== -1) {
+    throw new ERR_INVALID_MIME_SYNTAX('type', str, typeEnd);
   }
-  // skip type and '/'
+
+  // #6 Advance position by 1. (This skips past U+002F (/).)
   position = typeEnd + 1;
-  const type = toASCIILower(trimmedType);
-  // read until ';'
-  const subtypeEnd = StringPrototypeIndexOf(str, SEMICOLON, position);
-  const rawSubtype = subtypeEnd === -1 ?
-    StringPrototypeSlice(str, position) :
-    StringPrototypeSlice(str, position, subtypeEnd);
+
+  // #7 Let subtype be the result of collecting a sequence of code points
+  // that are not U+003B (;) from input, given position.
+  let subtypeEnd = StringPrototypeIndexOf(str, SEMICOLON, position);
+  subtypeEnd = subtypeEnd === -1 ? strLength : subtypeEnd;
```

Comment:
is branchless faster?

```suggestion
  subtypeEnd = subtypeEnd === -1 && strLength || subtypeEnd;
```

Reply:
strLength is not 0, so it should not skip to subtypeEnd